### PR TITLE
Add documentation about installation using Homebrew

### PR DESCRIPTION
**Pull request by stloyd**
<h2>Change Log</h2>
<div id="change-log">
  <h4>Added</h4>
  <ul id="added">
    <li>Add documentation about installation using Homebrew</li>
    <li>Add Homebrew formula into monorepo</li>
  </ul> 
  <h4>Fixed</h4>  
  <ul id="fixed">
    
  </ul>
  <h4>Changed</h4>
  <ul id="changed">
    
  </ul>  
  <h4>Removed</h4>
  <ul id="removed">
    
  </ul>
  <h4>Deprecated</h4>
  <ul id="deprecated">
    
  </ul>  
  <h4>Security</h4> 
  <ul id="security">
    
  </ul>     
</div>
<hr/>

<h2>Description</h2>

Example (commit message adjusted in PR):
Zrzut ekranu 2023-12-5 o 16 31 55

### CODE REVIEW

File: docs/installation/homebrew.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# Homebrew
+
+- [⬅️️ Back](../installation.md)
+
+To install box using [Homebrew](https://brew.sh), you need to tap the box formula first.
+
+```bash
+brew tap flow-php/flow
+brew install flow
```
Comment (norberttech) [inline docs/installation/homebrew.md]: I don't think this is actually installing flow-php, it seems it's installing that other `flow` tool available in the official distribution. 
To install our flow you would need to do `brew install flow-php/flow/flow`

File: .github/workflows/build-release.yml [hunk 2]
```diff
@@ -104,6 +104,15 @@ jobs:
...
           echo "BUILD_TAG=$BUILD_TAG" >> $GITHUB_ENV
 
+      - name: "Update Homebrew formula"
+        uses: dawidd6/action-homebrew-bump-formula@v3
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+          tap: flow-php/flow
+          formula: flow
+          tag: ${{ github.event.release.tag_name }}
+          revision: ${{ github.event.release.target_commitish }}
```
Comment (norberttech) [inline .github/workflows/build-release.yml]: Could you explain to me what this action is supposed to do?

Comment (stloyd) [reply .github/workflows/build-release.yml]: It should create PR with an update to the formula after release. But I will extract this into a new PR.

File: docs/installation/homebrew.md (hunk 1 above)
Comment (stloyd) [reply docs/installation/homebrew.md]: I have adjusted the formula name and now it's possible to run `brew install flow-php`:
```console
stloyd@MacBook-Pro-3 Downloads % brew tap flow-php/flow
==> Tapping flow-php/flow
Cloning into '/usr/local/Homebrew/Library/Taps/flow-php/homebrew-flow'...
remote: Enumerating objects: 21, done.
remote: Counting objects: 100% (21/21), done.
remote: Compressing objects: 100% (13/13), done.
remote: Total 21 (delta 5), reused 16 (delta 4), pack-reused 0
Receiving objects: 100% (21/21), 4.66 KiB | 2.33 MiB/s, done.
Resolving deltas: 100% (5/5), done.
Tapped 1 formula (15 files, 11.7KB).
stloyd@MacBook-Pro-3 Downloads % brew install flow-php
==> Downloading https://formulae.brew.sh/api/formula.jws.json
#################################################################################################################################################### 100.0%
==> Downloading https://formulae.brew.sh/api/cask.jws.json
#################################################################################################################################################### 100.0%
==> Fetching flow-php/flow/flow-php
==> Downloading https://github.com/flow-php/flow/releases/download/0.5.1/flow.phar
==> Downloading from https://objects.githubusercontent.com/github-production-release-asset-2e65be/370006659/af6c641a-2e3a-4286-b2a3-2f16c657bb4c?X-Amz-Algo
#################################################################################################################################################### 100.0%
==> Installing flow-php from flow-php/flow
🍺  /usr/local/Cellar/flow-php/0.5.1: 3 files, 8.6MB, built in 1 second

**Comment by norberttech** [comment]
I would prefer to keep everything code related in the monorepo, also I’m not sure if I understand the comment about GH_TOKEN

**Comment by stloyd** [comment]
Comment about GH token is outdated, it was related to the action for update of brew formula which is postponed to next PR.

**Comment by norberttech** [comment]
Did you check all automation on your fork first?

**Comment by norberttech** [comment]
Ok, so here is my position on this.

As much as I would love to keep it like that, there is one serious problem with the approach of keeping brew formula outside of the monorepo. Outside repo means another step in the release process, which I would prefer to keep as simple as possible. 

We have the following options here: 

- find a way to make it a part of the mono repo and automate the release process 
- open PR to brew official distro
- properly document the release process of ETL 

The official distro also means manually opening PR's after every release, but at least users wont need to add custom taps to brew. 
Personally, I would go with option number one, automating update of formula file after release build is done, should not be very hard.

File: .github/workflows/build-release.yml [hunk 3]
```diff
@@ -111,3 +111,22 @@ jobs:
...
+          formula-name: flow-php
+          formula-path: src/tools/homebrew/Formula/flow-php.rb
+          homebrew-tap: flow-php/flow
+          download-url: https://github.com/flow-php/flow/releases/download/${{ steps.get_version.outputs.version }}/flow.phar
+          commit-message: |
+            Update Homebrew TAP formula: {{formulaName}} to version: {{version}}
+
+            Created by https://github.com/mislav/bump-homebrew-formula-action
+        env:
+          COMMITTER_TOKEN: ${{ secrets.HOMEBREW_TAP_TOKEN }}
```
Comment (norberttech) [inline .github/workflows/build-release.yml]: can you rename into `${{ secrets.ACCESS_TOKEN }}` - its the same token automation is using for updating changelogs

Comment (norberttech) [reply .github/workflows/build-release.yml]: Its this one: 

image

File: .github/workflows/build-release.yml [hunk 4]
```diff
@@ -111,3 +111,22 @@ jobs:
           files: |
             build/flow.phar
             build/flow.phar.asc
+
+      - name: "Get version"
+        id: get_version
+        run: |
+          echo "version=${GITHUB_REF/refs\/tags\//}" >> $GITHUB_OUTPUT
```
Comment (norberttech) [inline .github/workflows/build-release.yml]: I think you should be able to use `$BUILD_TAG` that is extracted one step before this one.

File: src/tools/homebrew/Formula/flow-php.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
+class FlowPhp < Formula
+  desc "Flow is a first and the most advanced PHP ETL framework"
+  homepage "https://github.com/flow-php/flow"
+  url "https://github.com/flow-php/flow/releases/download/0.5.1/flow.phar"
+  sha256 "a96caaccf4382183edbd8a43e41b4c6307bba717867480aa69711ade7bca7204"
+
+  depends_on "php" if MacOS.version <= :el_capitan
```
Comment (norberttech) [inline src/tools/homebrew/Formula/flow-php.rb]: Why does it depend on php only when a version is lower or equal to el_capitan?

File: .github/workflows/build-release.yml (hunk 4 above)
Comment (stloyd) [reply .github/workflows/build-release.yml]: Tag doesn't contain dots, which leads to broken URL:
```console
Run mislav/bump-homebrew-formula-action@v3
  with:
    formula-name: flow-php
    formula-path: src/tools/homebrew/Formula/flow-php.rb
    homebrew-tap: stloyd/flow
    base-branch: feature/homebrew
    download-url: https://github.com/stloyd/flow/releases/download/054/flow.phar
    commit-message: {{formulaName}} {{version}}
  
  Created by https://github.com/mislav/bump-homebrew-formula-action
  
  env:
    COMPOSER_PROCESS_TIMEOUT: 0
    COMPOSER_NO_INTERACTION: 1
    COMPOSER_NO_AUDIT: 1
    BUILD_TAG: 054
    COMMITTER_TOKEN: ***
GET /repos/stloyd/flow/releases/tags/054 - 404 in 311ms
```

Comment (norberttech) [reply .github/workflows/build-release.yml]: In this case, I would move definitions of those two variables to the beginning of the workflow and add some comments about them.

Comment (stloyd) [reply .github/workflows/build-release.yml]: Fixed that build tag (renamed it too) as it's some leftover from past, and used in new workflow.

File: src/tools/homebrew/Formula/flow-php.rb (hunk 5 above)
Comment (stloyd) [reply src/tools/homebrew/Formula/flow-php.rb]: Outdated, should be always php.

**Comment by norberttech** [comment]
Could you, just one last time, on your fork, mark a new release and test this whole build-release workflow before I merge this and release new version?

**Comment by stloyd** [comment]
@norberttech Here:
- the commit: https://github.com/stloyd/flow/commit/32c9bc51e8760ab6e5e24930416cf5df3c6a55bd
- the release action: https://github.com/stloyd/flow/actions/runs/7104378394/job/19339754055

On my fork the main difference was old comment + usage of my fork in all URLs.

When you approve this PR I will rebase and remove the commit from the PR.

**Comment by norberttech** [comment]
Looks good, any idea why on hour 0.5.6 version there is no signed phar? 
Also, are there any concerns about the repo split? Will that additional commit trigger another split after marking a new release?

**Comment by stloyd** [comment]
1. I have disabled sign of phar as I didn't setup it on my fork, it will not affect anything as it uses the phar file only,
2. Split should be executed normally as this would be a new commit to `1.x` branch, so it should be propagated to the `flow-php/homebrew-flow` repository

**Comment by norberttech** [comment]
yeah, but what about the version of the tap repo? Could you maybe explain step by step what will happen once release is published? I mean which workflows will
be executed and in what order?

**Comment by stloyd** [comment]
1. Release trigger will start build workflow,
2. Build workflow does phar, uploads it on release tag artifact, at end will trigger new commit to main branch with update of formula file with details of new phar,
3. New commit will trigger split on mono-repository pointing to `homebrew-flow` as above commit will land in that package,
4. Done.

The only consideration here can be changelog update being triggered on commit to update formula. Probably we should add that path to ignore, like:
```diff
Subject: [PATCH] Add homebrew formula into monorepo
---
Index: .github/workflows/automation-changelog-update.yml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.github/workflows/automation-changelog-update.yml b/.github/workflows/automation-changelog-update.yml
--- a/.github/workflows/automation-changelog-update.yml	(revision 8e9db5ec94ef1f2262485104d3160eec2f0282d4)
+++ b/.github/workflows/automation-changelog-update.yml	(date 1701802523024)
@@ -4,6 +4,9 @@
   workflow_dispatch:
   push:
     branches: [ 1.x ]
+    paths-ignore:
+      # Ignore updates of Homebrew formula in CHANGELOG
+      - 'src/tools/homebrew/Formula/flow-php.rb'
 
 jobs:
   automation:
```

**Comment by norberttech** [comment]
> The only consideration here can be changelog update being triggered on commit to update formula. Probably we should add that path to ignore, like:

I think that's fine. It will be a single entry after each release.

**Review by norberttech — APPROVED** [review]

**Comment by norberttech** [comment]
Ping me when it's ready to merge, will push another release after

**Pull request merged**